extensions/exchanges/bitfinex/exchange.js

```diff
@@ -56,201 +248,163 @@ module.exports = function container (get, set, clear) {
+        var args = {}
+        args.sort = -1 //backward
+        args.limit = 1000
+        if (opts.from) {
+          args.start = opts.from
+        }
+        else if (opts.to) {
+          args.end = opts.to
+        }
+        else if (args.start && !args.end) {
+          args.end = args.start + 500000
+        }
+        else if (args.end && !args.start) {
+          args.start = args.end - 500000
+        }
+        var query = encodeQueryData(args)
+        var tpair = 't' + joinProduct(opts.product_id)
+        client.makePublicRequest('trades/' + tpair + '/hist?' + query, function (err, body) {
+          if (err) return retry('getTrades', opts, cb)
+          var trades = body.map(function(trade) {
+            return {
+              trade_id: trade.ID,
+              time: trade.MTS,
+              size: Math.abs(trade.AMOUNT),
+              price: trade.PRICE,
+              side: trade.AMOUNT > 0 ? 'buy' : 'sell'
+            }
+          })
+          cb(null, trades)
+        }) 
+      } else {
+        // We're live now (i.e. opts.from is set), use websockets
+        if (typeof(ws_trades) === "undefined") { return retry('getTrades', opts, cb) }
+        trades = ws_trades.filter(function (trade) { return trade.time >= opts.from })
         cb(null, trades)
-      })
+      }
     },
     
     getBalance: function (opts, cb) {
-      var client = authedClient()
-      client.wallet_balances(function (err, body) {
-        if (err) return(err)
-        var balance = {asset: 0, currency: 0}
-        var accounts = _(body).filter(function (body) { return body.type === c.bitfinex.wallet }).forEach(function (account) {
-          if (account.currency.toUpperCase() === opts.currency) {
-            balance.currency = n(account.amount).format('0.00000000')
-            balance.currency_hold = n(account.amount).subtract(account.available).format('0.00000000')
-          }
-          else if (account.currency.toUpperCase() === opts.asset) {
-            balance.asset = n(account.amount).format('0.00000000')
-            balance.asset_hold = n(account.amount).subtract(account.available).format('0.00000000')
-          }
-        })
-        cb(null, balance)
-      })
+      if (!authed_client_ws) { authedClientWs() }
+      if (Object.keys(ws_balance).length === 0) { return retry('getBalance', opts, cb) }
+
+      balance = {}
+      balance.currency      = n(ws_balance[opts.currency].balance).format('0.00000000')
+      balance.asset         = n(ws_balance[opts.asset].balance).format('0.00000000')
+
+      balance.currency_hold = ws_balance[opts.currency].available ? n(ws_balance[opts.currency].balance).subtract(ws_balance[opts.currency].available).format('0.00000000') : n(0).format('0.00000000')
+      balance.asset_hold    = ws_balance[opts.asset].available    ? n(ws_balance[opts.asset].balance).subtract(ws_balance[opts.asset].available).format('0.00000000')       : n(0).format('0.00000000')
+
+      cb(null, balance)
```

Comment:
Plain example:
```
---------------------------- STARTING LIVE TRADING ----------------------------

Waiting 1s for initial websockets snapshot.

Balance WS: { currency: '0.00000000',
  asset: '0.04180751',
  currency_hold: '0.00000000',
  asset_hold: '0.00000000' }

Balance WS: { currency: '0.00000000',
  asset: '0.04180751',
  currency_hold: '0.00000000',
  asset_hold: '0.00000000' }
2017-08-09 17:34:47  3308.40 BTC-USD   +0.1%        6   61 ++      0.0158   0.0099           0.04181 BTC  0.00000 USD    +1.8%   +0.0%
```
```
---------------------------- STARTING LIVE TRADING ----------------------------

Balance REST: { asset: '0.04180751',
  currency: '0.00000000',
  asset_hold: '0.03265206',
  currency_hold: '0.00000000' }

Balance REST: { asset: '0.04180751',
  currency: '0.00000000',
  asset_hold: '0.03265206',
  currency_hold: '0.00000000' }
2017-08-09 17:37:24  3310.00 BTC-USD   +0.0%       13   63 ++      0.0180   0.0092           0.04181 BTC  0.00000 USD    +1.8%   +0.0%
```
and this it not so tragic as **this**:

Should be like that:
```
2017-08-09 17:38:47  3310.10 BTC-USD   +0.0%        6   63 ++      0.0169   0.0091           0.04181 BTC  0.00000 USD    +1.9%   +0.0%
2017-08-09 17:38:52 - sell delayed: +78.1% of funds (0.03265206 BTC) on hold

Balance REST: { asset: '0.04180751',
  currency: '0.00000000',
  asset_hold: '0.03265206',
  currency_hold: '0.00000000' }
```
Is like this:
```
2017-08-09 17:39:41  3310.10 BTC-USD   +0.0%       26   63 ++      0.0169   0.0091           0.04181 BTC  0.00000 USD    +1.9%   +0.0%
2017-08-09 17:39:58 - placing sell order...
2017-08-09 17:39:58 - sell order placed at 3310.10 USD

Balance WS: { currency: '0.00000000',
  asset: '0.04180751',
  currency_hold: '0.00000000',
  asset_hold: '0.00000000' }

2017-08-09 17:40:03 - order status: rejected
2017-08-09 17:40:03 - not enough balance for sell, aborting
2017-08-09 17:40:03 - not enough balance, or signal switched, cancel sell
```